-- local fs = require("@lune/fs")
-- local stdio = require("@lune/stdio")

-- local clappy = require("@vendor/clappy")

-- local constants = require("../constants")
-- local fmt = require("../fmt")

-- local function run(_: any, opts: { force: boolean? })

-- end

-- return clappy.command({
-- 	name = "run",
-- 	description = "Initializes the current folder with a Bright setup",

-- 	parameters = {
-- 		transformers = clappy.parameter({
-- 			position = 1,
-- 			description = "The transformers to execute",
-- 			type = clappy.types.
-- 		})
-- 	},
-- 	opts = {
-- 		force = clappy.opt({
-- 			description = "Forcefully overwrite any Bright configuration non-interactively",
-- 			type = clappy.types.boolean(),
-- 		}),
-- 	},
-- }, run)

return nil
